### PR TITLE
Fix TS lobby map previews

### PR DESCRIPTION
**Pull request by pchote**
previews

### CODE REVIEW

File: OpenRA.Game/Graphics/Minimap.cs [hunk 1]
```diff
@@ -70,23 +98,29 @@ static Bitmap AddStaticResources(TileSet tileset, Map map, Ruleset resourceRules
...
 
-						colors[y * stride + x] = tileset[tileset.GetTerrainIndex(res)].Color.ToArgb();
+						var color = tileset[tileset.GetTerrainIndex(res)].Color.ToArgb();
+						if (isDiamond)
+						{
+							// Odd rows are shifted right by 1px
+							var dx = uv.V & 1;
+							if (x + dx > 0)
+								colors[y * stride + 2*x + dx - 1] = color;
+							colors[y * stride + 2*x + dx] = color;
```
Comment (Phrohdoh) [inline OpenRA.Game/Graphics/Minimap.cs]: Newline above this, please.

File: OpenRA.Game/Graphics/Minimap.cs [hunk 2]
```diff
@@ -38,16 +47,25 @@ public static Bitmap TerrainBitmap(TileSet tileset, Map map, bool actualSize = f
...
+						var leftColor = type != null ? type.LeftColor : Color.Black;
+
+						if (isDiamond)
+						{
+							// Odd rows are shifted right by 1px
+							var dx = uv.V & 1;
+							var rightColor = type != null ? type.RightColor : Color.Black;
+							if (x + dx > 0)
+								colors[y * stride + 2*x + dx - 1] = leftColor.ToArgb();
+							colors[y * stride + 2*x + dx] = rightColor.ToArgb();
```
Comment (Phrohdoh) [inline OpenRA.Game/Graphics/Minimap.cs]: Newline above this too, please.

**Comment by penev92** [comment]
Code looks OK and map previews look very good. I was going to +1 until the weirdest thing happened.
I have the `blank-conquest` map that was removed a while back. Apparently generating a preview for it is too hard, because it crashes the game every single time, with a variety of exceptions that all boil down to `OutOfMemory` (either `System.OutOfMemoryException` or an OpenGL `OutOfMemory`, with the occasional "attempt to write in protected memory" or something along those lines).
Important notes:
 - This map works fine on bleed!
 - This is the only one of the available maps that crashes the game (although the others are severely bigger, they don't cause crashes).
 - What is even better is that I don't even get logs for this.

**Comment by pchote** [comment]
Works for me.  Are you sure your map just isn't bogus?  We now have stricter requirements about valid map bounds.
screen shot 2015-07-29 at 08 13 17

**Comment by MicroJOo** [comment]
i have created a new map in the editor ... looks like this : 

preview1

but when i try to load the map , the preview of it looks like this : 

preview2

**Comment by penev92** [comment]
@MicroJOo the radar widget is the next thing to be fixed.
You can see if the preview is correct by the map size - 128x144. Since tiles are twice as wide as they are high, it becomes 256x144, which looks like the map preview.

@pchote well can you send me the map then? Like I said mine works on bleed, but...

**Comment by penev92** [comment]
:+1:

File: OpenRA.Game/Graphics/Minimap.cs [hunk 3]
```diff
@@ -70,23 +99,30 @@ static Bitmap AddStaticResources(TileSet tileset, Map map, Ruleset resourceRules
...
+					for (var x = 0; x < width; x++)
 					{
-						var mapX = x + b.Left;
-						var mapY = y + b.Top;
-						if (map.MapResources.Value[new MPos(mapX, mapY)].Type == 0)
+						var uv = new MPos(x + b.Left, y + b.Top);
+						if (map.MapResources.Value[uv].Type == 0)
 							continue;
 
-						var res = resourceRules.Actors["world"].Traits.WithInterface<ResourceTypeInfo>()
```
Comment (Mailaender) [inline OpenRA.Game/Graphics/Minimap.cs]: Ugh, this trait lookup was done for every pixel in the minimap. Nice cleanup. :+1:

**Comment by Mailaender** [comment]
Also can't spot any regressions for non-isometric maps. :white_check_mark:

**Comment by obrakmann** [comment]
Corrects the ts map previews, the other mods still work, too. :+1:

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/a4adabed94666ae819ab17bdc9aa378e220b9ae7%5E...a4adabed94666ae819ab17bdc9aa378e220b9ae7)

**Pull request merged**